components/director/internal/domain/formationconstraint/constraint_engine.go

```diff
@@ -36,28 +36,50 @@ type labelRepository interface {
 	GetByKey(ctx context.Context, tenant string, objectType model.LabelableObject, objectID, key string) (*model.Label, error)
 }
 
+//go:generate mockery --exported --name=applicationRepository --output=automock --outpkg=automock --case=underscore --disable-version-string
+type applicationRepository interface {
+	ListByScenariosNoPaging(ctx context.Context, tenant string, scenarios []string) ([]*model.Application, error)
+}
+
+//go:generate mockery --exported --name=labelService --output=automock --outpkg=automock --case=underscore --disable-version-string
+type labelService interface {
+	GetByKey(ctx context.Context, tenant string, objectType model.LabelableObject, objectID, key string) (*model.Label, error)
+}
+
 // ConstraintEngine determines which constraints are applicable to the reached join point and enforces them
 type ConstraintEngine struct {
 	constraintSvc             formationConstraintSvc
 	tenantSvc                 tenantService
 	asaSvc                    automaticScenarioAssignmentService
 	formationRepo             formationRepository
 	labelRepo                 labelRepository
+	labelService              labelService
+	applicationRepository     applicationRepository
 	operators                 map[OperatorName]OperatorFunc
 	operatorInputConstructors map[OperatorName]OperatorInputConstructor
+	applicationTypeLabelKey   string
```

Comment:
We discussed offline whether `applicationTypeLabelKey` to be passed in the constructor or to be added in the constraint inputTemplate -> joinPointDetails -> operatorInput.

As I gave it a further thought I lean more towards the second option for the following 3 reasons:
- if we can calculate a field that will be used by the engine in the joinPointDetails, why not to do it there but separate the places from where 'input' information is passed; if we move the key to the joinPointDetails we will encapsulate the logic in one place;
- if we add the `applicationTypeLabelKey` to joinPointDetails (in our case `AssignFormationOperationDetails`) we will know more granually in which flow this field is needed/will be used whereas if we add it directly in the constructor, initially we won't have an idea why and where this field will be used. In the future, if more such fields are added, from my PoV it will become messy.
- if we add it in the joinPointDetails, the resource type key will be determined dynamically based on the object type that is passed, meaning that we will have it out-of-the-box for all supported object types by `AssignFormation`. If we leave it in the constructor and in the future we will have to use the runtimeType key, we will have to add it separately as another argument and so on for every other type.

What do you think, I am open for a discussion.